File: 240616/정수 사각형 차이의 최소 2/minimum-difference-on-the-integer-grid-2.py
n = int(input())
board = [list(map(int, input().split())) for _ in range(n)]

dp = [[[-int(1e9), int(1e9)]] * n for _ in range(n)]

dp[0][0] = [board[0][0], board[0][0]]

for x in range(1, n):
    v = board[0][x]
    dp[0][x] = [max(dp[0][x-1][0], v), min(dp[0][x-1][1], v)]

for y in range(1, n):
    v = board[y][0]
    dp[y][0] = [max(dp[y-1][0][0], v), min(dp[y-1][0][1], v)]

for y in range(1, n):
    for x in range(1, n):
        v = board[y][x]
        a1, a2= dp[y-1][x]
        b1, b2 = dp[y][x-1]

        a1, a2 = max(a1, v), min(a2, v)
        b1, b2 = max(b1, v), min(b2, v)

        if a1 - a2 < b1 - b2:
            dp[y][x] = [a1, a2]
        else:
            dp[y][x] = [b1, b2]

print(dp[n-1][n-1][0] - dp[n-1][n-1][1])

File: 241014/정수 사각형 차이의 최소 2/minimum-difference-on-the-integer-grid-2.java
import java.util.*;
public class Main {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int N = sc.nextInt();
        int[][] board = new int[N][N];
        int[][][] dp = new int[N][N][2];
        for(int i = 0; i < N; i++){
            for(int j = 0; j < N; j++){
                board[i][j] = sc.nextInt();
            }
        }
        dp[0][0][0] = board[0][0];
        dp[0][0][1] = board[0][0];
        for(int i = 1; i < N; i++){
            dp[0][i][0] = Math.min(dp[0][i-1][0], board[0][i]);
            dp[0][i][1] = Math.max(dp[0][i-1][1], board[0][i]);
        }

        for(int i = 1; i < N; i++){
            dp[i][0][0] = Math.min(dp[i-1][0][0], board[i][0]);
            dp[i][0][1] = Math.max(dp[i-1][0][1], board[i][0]);
        }

        for(int i = 1; i < N; i++){
            for(int j = 1; j < N; j++){
                int x1 = Math.min(board[i][j], dp[i-1][j][0]);
                int y1 = Math.max(board[i][j], dp[i-1][j][1]);

                int x2 = Math.min(board[i][j], dp[i][j-1][0]);
                int y2 = Math.max(board[i][j], dp[i][j-1][1]);

                int a = Math.abs(x1 - y1);
                int b = Math.abs(x2 - y2);

                if( a < b){
                    dp[i][j][0] = x1;
                    dp[i][j][1] = y1;
                }else{
                    dp[i][j][0] = x2;
                    dp[i][j][1] = y2;                    
                }
            }
        }

        int ans = Math.abs(dp[N-1][N-1][0] - dp[N-1][N-1][1]);
        System.out.println(ans);
    }
}